test/jdk/tools/jlink/JLinkReproducible3Test.java

```diff
@@ -0,0 +1,112 @@
+    public static void main(String[] args) throws Exception {
+        Path image1 = Paths.get("./image1");
+        Path image2 = Paths.get("./image2");
+
+        Path copy_jdk1_dir = Path.of("./copy-jdk1-tmpdir");
+        Path copy_jdk2_dir = Path.of("./copy-jdk2-tmpdir");
+        Path jdk_test_dir = Path.of(
+                Optional.of(
+                        System.getProperty("test.jdk"))
+                        .orElseThrow(() -> new RuntimeException("Couldn't load JDK Test Dir"))
+        );
+
+        copyJDKs(jdk_test_dir, copy_jdk1_dir, copy_jdk2_dir);
+
+        Path copied_jlink1 = Optional.of(
+                Paths.get(copy_jdk1_dir.toString(), "bin", "jlink"))
+                .orElseThrow(() -> new RuntimeException("Unable to load copied jlink")
+                );
+
+        Path copied_jlink2 = Optional.of(
+                Paths.get(copy_jdk2_dir.toString(), "bin", "jlink"))
+                .orElseThrow(() -> new RuntimeException("Unable to load copied jlink")
+                );
+
+        runCopiedJlink(copied_jlink1.toString(), "--add-modules", "java.base,jdk.management,jdk.unsupported,jdk.charsets", "--output", image1.toString());
+        runCopiedJlink(copied_jlink2.toString(), "--add-modules", "java.base,jdk.management,jdk.unsupported,jdk.charsets", "--output", image2.toString());
+
+        long mismatch = Files.mismatch(image1.resolve("lib").resolve("modules"), image2.resolve("lib").resolve("modules"));
+        if (mismatch != -1L) {
+            throw new RuntimeException("jlink producing inconsistent result in modules. Mismatch in modules file occurred at byte position " + mismatch);
+        }
+    }
+
+    private static void runCopiedJlink(String... args) throws Exception {
+        var pb = new ProcessBuilder(args);
+        var res = ProcessTools.executeProcess(pb);
+        res.shouldHaveExitValue(0);
+    }
+
+    private static String runJavaVersion(Path jdk_test_base_dir) throws Exception {
+        var java_exec = Paths.get(jdk_test_base_dir.toString(), "bin", "java");
+        var pb = new ProcessBuilder(java_exec.toString(), "--version");
+        var res = ProcessTools.executeProcess(pb);
+        res.shouldHaveExitValue(0);
+        return res.getStdout();
+
+
+    }
+
+    private static void copyJDKs(Path src, Path dst1, Path dst2) throws Exception {
+        Files.walk(src).skip(1).forEach(file -> {
+            try {
+                System.out.println(file);
+                if(Files.isDirectory(file)) {
+                    Files.createDirectories(dst1.resolve(src.relativize(file)));
+                    Files.createDirectories(dst2.resolve(src.relativize(file)));
+                } else {
+                    Files.copy(file, dst1.resolve(src.relativize(file)));
+                    Files.copy(file, dst2.resolve(src.relativize(file)));
+                }
+            } catch (IOException ioe) {
+                throw new UncheckedIOException(ioe);
+            }
+        });
+    }
+}
+
```

Comment:
The update to ImageFileCreator looks good.

A few nits in JLinkReproducible3Test:
1. Invoking copyJDKs with two destination locations looks a bit strange. I think would be easier for future maintainers if it does one copy and is called twice.
2. Files.copy will create a directory, no need for Files.createDirectories.
3. Are you sure that bin/jlink is executable, I just wonder if the test runs with the patch have ended up depending the umask at the time. It is possible to specify COPY_ATTRIBUTES to the copy method and it will attempt to copy the permissions.
4. The variable naming is more like C style, so a bit inconsistent with the naming usually used in Java programs.

Reply:
1. I agree that it's awkward and I'm not a huge fan. I'm okay doing it this way, but this is recreating the reproducer for https://bugs.openjdk.java.net/browse/JDK-8252730 which is two identical JDK's in different directories. I can document this in the test further or we can change it.

2-4: Noted. Will clean these up and loop back.

Reply:
Whatever is easiest but 2 x copy tree would be simpler and the result should be two identical JDKs as reported in JDK-8252730.

Reply:
I misunderstood what you meant when I read it earlier today, apologies! The new changes should address these enumerated issues.

Reply:
The updated test looks okay, I can sponsor if you need it.